files/tk_metrics

```diff
@@ -7,44 +7,69 @@ require 'time'
-end.parse!
+#===========================================================================#
+# BUILD CONFIGURATION OPTIONS                                               #
+#===========================================================================#
 
-if options[:metrics_type].nil? then
-  STDERR.puts '--metrics_type (-m) is a required argument'
-  exit 1
-end
+settings = { }
+config = { }
+
+OptionParser.new do |parser|
+  parser.banner = "Usage: tk_metrics [options]"
+
+  save_ident = lambda { |name,arg| config[name] = arg }
+  save_list  = lambda { |name,arg| config[name] = arg.split(',') }
 
-METRICS_TYPE = options[:metrics_type]
-config = YAML.load_file(File.join(File.dirname(File.expand_path(__FILE__)),"#{METRICS_TYPE}_config.yaml"))
+  opt_bool  = lambda { |arg| "--[no-]#{arg}" }
+  opt_value = lambda { |arg| "--#{arg} ARG" }
 
-def coalesce ( opt1, opt2, default=nil )
-  if opt1.nil? then
-    if opt2.nil? then
-      default
-    else
-      opt2
+  option = lambda do |name,desc,fopt,fsave,default=nil|
+    settings[name] = { default: default }
+    parser.on(fopt.call(name.to_s), desc) do |arg|
+      fsave.call(name, arg)
     end
-  else
-    opt1
   end
+
+  # Each possible configuration option is defined here
+  option.call(:"output-dir", 'Directory to save output to', opt_value, save_ident, false)
+  option.call(:hosts, 'Hosts to collect metrics from (comma-separated)', opt_value, save_list, ['localhost'])
+  option.call(:"additional-metrics", 'Additional metrics to collect (comma-separated)', opt_value, save_list, [])
+  option.call(:"metrics-port", 'The port the metrics service runs on', opt_value, save_ident)
+  option.call(:clientcert, 'Not used', opt_value, save_ident)
+  option.call(:"pe-version", 'The version of PE in use', opt_value, save_ident)
+  option.call(:print, 'Print to stdout', opt_bool, save_ident, true)
+  option.call(:ssl, 'Whether or not to use SSL when gather metrics', opt_bool, save_ident, true)
+
+  # This one is special because it's required
+  option.call(:"metrics-type", 'Type of metric to collect', opt_value, save_ident)
+
+  # Kinda pointless to include descriptions unless there's a help flag
+  parser.on("-h", "--help", "Prints this help") { puts parser; exit }
+
+end.parse!
+
+raise('ERROR: Missing configuration option "metrics-type"') if config[:"metrics-type"].nil?
+
+begin
+  file = File.join(File.dirname(File.expand_path(__FILE__)),"#{config[:"metrics-type"]}_config.yaml")
+  file_config = YAML.load_file(file)
+rescue Exception => e
+  STDERR.puts "ERROR: unable to read configuration file from #{file}: #{e}"
+  exit 1
+end
+
+# For the options which MAY be read from a config file, retrieve those values
+# if they have not been specified during invocation. Values specified on the
+# command line are given precedence.
+settings.each do |opt,attrs|
+  possibilities = [config[opt], file_config[opt], attrs[:default]]
+  config[opt] = possibilities.find {|p| !p.nil? }
```

Comment:
Thoughts on reading in the config file before parsing command line args.  Then you can check
 
if the command line arg is nil
  use the config file setting 
  if the config file setting is nil use the default
  else fail